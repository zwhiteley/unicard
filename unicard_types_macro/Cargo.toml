[package]
name = "unicard_types_macro"
version = "0.1.0"
authors = ["Zachary Whiteley <z.n.whiteley@gmail.com>"]
description = "Traits and structs for reading/writing to WASM linear memory"
repository = "https://github.com/zwhiteley/unicard"
license = "MIT OR Apache-2.0"
edition = "2021"

[lib]
proc-macro = true

[dependencies]
syn = "2.0.38"
quote = "1.0.33"
proc-macro2 = "1.0.69"

[dev-dependencies]
trybuild = "1.0.85"
unicard_types = { path = "../unicard_types" }
