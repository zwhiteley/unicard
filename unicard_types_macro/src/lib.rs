use proc_macro2::TokenStream;
use quote::quote_spanned;
use syn::spanned::Spanned;
use syn::{parse_macro_input, Data, DeriveInput};

mod shared;

mod struct_derive;

mod enum_derive;

#[proc_macro_derive(WasmType)]
pub fn wasm_type_macro(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
    let input = parse_macro_input!(input as DeriveInput);
    proc_macro::TokenStream::from(wasm_type(input))
}

fn wasm_type(input: DeriveInput) -> TokenStream {
    // There is little need for generic support, so don't implement it. This includes
    // support for lifetimes -- as `WasmType::read` allows for a value to be read,
    // it is impossible for any meaningful lifetime parameters to bee included
    if !input.generics.params.is_empty() {
        return quote_spanned!(
            input.generics.span() => ::std::compile_error!("WasmType does not support generics");
        );
    }

    // NOTE: as this is an internal macro, we do not need to support crate renaming
    match input.data {
        Data::Struct(data_struct) => struct_derive::derive(input.ident, data_struct),
        Data::Enum(data_enum) => enum_derive::derive(input.ident, data_enum),
        Data::Union(_) => {
            let span = proc_macro2::Span::call_site();
            quote_spanned!(span => ::std::compile_error!("the WasmType macro cannot be derived for unions");)
        }
    }
}
