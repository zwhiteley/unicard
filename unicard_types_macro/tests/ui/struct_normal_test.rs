pub use unicard_types::*;
use unicard_types_macro::WasmType32 as WasmType;

#[derive(WasmType)]
struct Test {
    // NOTE: pointers will never be supported due to the lack of dereferencability
    // guarantees and an inability to pass them directly to WASM safely (except as
    // opaque types, although few languages support the use of such a feature).
    inner: *const i32
}

fn main() {}