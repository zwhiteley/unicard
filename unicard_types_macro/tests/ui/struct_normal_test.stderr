error[E0277]: the trait bound `*const i32: unicard_types::WasmType32` is not satisfied
 --> tests/ui/struct_normal_test.rs:9:12
  |
9 |     inner: *const i32
  |            ^^^^^^^^^^ the trait `unicard_types::WasmType32` is not implemented for `*const i32`
  |
  = help: the trait `unicard_types::WasmType32` is implemented for `i32`
